firestore/google/cloud/firestore_v1beta1/_helpers.py

```diff
@@ -119,6 +120,56 @@ def __ne__(self, other):
             return not equality_val
 
 
+class FieldPath(object):
+    """ Field Path object for client use.
+
+    Args:
+        parts (string or iterable): Indicating path of the key to be used.
+    """
+
+    def __init__(self, *parts):
+        invalid_characters = '~*/[]'
+        if len(parts) == 1:
+            parts = parts[0]
+        if isinstance(parts, basestring):
+            parts = tuple(parts.split('.'))
+        for part in parts:
+            if (not part
+                or not isinstance(part, basestring)
+                or part in invalid_characters):
```

Comment:
This limitation only applies when the field path is supplied using a string via any of the API method that consume these paths.

`query.where('~', '==', 'foo')` is invalid, but `query.where(new FieldPath('~'), '==', 'foo')` is not.

Reply:
This confusion goes away if `FieldPath` always constructs a field path from component strings.

Reply:
I actually missed this in my review. I think what you want is a static method `from_string`, so that `FieldPath.from_string` checks for invalid characters and splits on dots, then constructs a `FieldPath`. Call that in `update`, `query`, etc. after doing an `isinstance` to distingish a string from a `FieldPath`. All the internal logic can then work solely with `FieldPath`s.